File: jlpt/native_ios/Sources/Data/API/ContentAPI.swift
import Foundation

final class ContentAPI {
    private let client: APIClient
    init(client: APIClient) { self.client = client }

    func getVocabulary(level: String, page: Int, size: Int) async throws -> PageDTO<VocabularyResponseDTO> {
        try await client.send(Endpoint(path: "api/vocabulary", query: pageQuery(level: level, page: page, size: size)))
    }

    func getKanji(level: String, page: Int, size: Int) async throws -> PageDTO<KanjiResponseDTO> {
        try await client.send(Endpoint(path: "api/kanji", query: pageQuery(level: level, page: page, size: size)))
    }

    func getGrammar(level: String, page: Int, size: Int) async throws -> PageDTO<GrammarResponseDTO> {
        try await client.send(Endpoint(path: "api/grammar", query: pageQuery(level: level, page: page, size: size)))
    }

    func getPassage(id: Int64) async throws -> PassageResponseDTO {
        try await client.send(Endpoint(path: "api/passages/\(id)"))
    }

    func getListeningAudio(id: Int64) async throws -> ListeningAudioResponseDTO {
        try await client.send(Endpoint(path: "api/listening-audios/\(id)"))
    }

    private func pageQuery(level: String, page: Int, size: Int) -> [URLQueryItem] {
        [
            URLQueryItem(name: "level", value: level),
            URLQueryItem(name: "page", value: String(page)),
            URLQueryItem(name: "size", value: String(size))
        ]
    }
}

File: jlpt/native_android/app/src/main/java/com/jlpt/android/data/remote/api/ContentApi.kt
package com.jlpt.android.data.remote.api

import com.jlpt.android.data.remote.dto.GrammarResponseDto
import com.jlpt.android.data.remote.dto.KanjiResponseDto
import com.jlpt.android.data.remote.dto.ListeningAudioResponseDto
import com.jlpt.android.data.remote.dto.PageDto
import com.jlpt.android.data.remote.dto.PassageResponseDto
import com.jlpt.android.data.remote.dto.VocabularyResponseDto
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query

interface ContentApi {
    @GET("api/vocabulary")
    suspend fun getVocabulary(
        @Query("level") level: String,
        @Query("page") page: Int,
        @Query("size") size: Int
    ): PageDto<VocabularyResponseDto>

    @GET("api/kanji")
    suspend fun getKanji(
        @Query("level") level: String,
        @Query("page") page: Int,
        @Query("size") size: Int
    ): PageDto<KanjiResponseDto>

    @GET("api/grammar")
    suspend fun getGrammar(
        @Query("level") level: String,
        @Query("page") page: Int,
        @Query("size") size: Int
    ): PageDto<GrammarResponseDto>

    @GET("api/passages/{id}")
    suspend fun getPassage(@Path("id") id: Long): PassageResponseDto

    @GET("api/listening-audios/{id}")
    suspend fun getListeningAudio(@Path("id") id: Long): ListeningAudioResponseDto
}
